Rails 7.2

jrafanie opened the pull request:
Fixes #23201 

TODO:

- [x] Release virtual attributes and use it, 🙇 Keenan: https://github.com/ManageIQ/activerecord-virtual_attributes/pull/187
- [x] Fix credentials in provider repos, 🙇 Fryguy: https://github.com/ManageIQ/manageiq/pull/23496
- [x] 💚 on [cross repo](https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/971)
- [x] Test with rails 7.2 defaults
- [x] Set rails 7.2 minimum to latest before merge


Details:

Here's the load_defaults for 7.2: https://github.com/rails/rails/blob/v7.2.2.1/railties/lib/rails/application/configuration.rb#L321-L337


```ruby
        when "7.2"
          load_defaults "7.1"

          self.yjit = true

          if respond_to?(:active_job)
            active_job.enqueue_after_transaction_commit = :default
          end

          if respond_to?(:active_storage)
            active_storage.web_image_content_types = %w( image/png image/jpeg image/gif image/webp )
          end

          if respond_to?(:active_record)
            active_record.postgresql_adapter_decode_dates = true
            active_record.validate_migration_timestamps = true
          end
```


* postgresql_adapter_decode_dates: https://www.github.com/rails/rails/pull/51483 https://www.github.com/rails/rails/pull/51763
* validate_migration_timestamps: https://www.github.com/rails/rails/pull/50400
* yjit: https://www.github.com/rails/rails/pull/51894

Fryguy commented:
I saw this rubocop warning elsewhere - I'm not sure exactly what's causing it, but I think we need to upgrade rubocop (and by extension that means dropping codeclimate).

kbrock commented on Gemfile:
```diff
@@ -21,7 +21,7 @@ manageiq_plugin "manageiq-schema"
 # Unmodified gems
 gem "activerecord-session_store",       "~>2.0"
-gem "activerecord-virtual_attributes",  "~>7.1.0"
+gem "activerecord-virtual_attributes",  "~>7.1.1"
```
the goal is to release 7.2, since that will be rails 7.2 friendly. That should fix up a lot of stuff.

We were thinking we may want `~>7.2.0.0` to give us the flexibility to upgrade to newer features. This may be overkill and we can discuss.

kbrock commented:
Surprised we didn't get a Gemfile issue around a rails conflict.
We need to merge https://github.com/ManageIQ/activerecord-virtual_attributes/pull/187

jrafanie commented:
> Surprised we didn't get a Gemfile issue around a rails conflict.
> We need to merge [ManageIQ/activerecord-virtual_attributes#187](https://github.com/ManageIQ/activerecord-virtual_attributes/pull/187)

We currently allow rails ~> 7.1 in VA as long as it's greater than 7.1.x.

Just pushed a commit to test against the v72 branch on @kbrock's fork of VA.

Fryguy commented on Gemfile:
```diff
@@ -69,7 +69,7 @@ gem "psych",                            ">=3.1",             :require => false #
 gem "rack",                             ">=2.2.14",          :require => false
 gem "rack-attack",                      "~>6.5.0",           :require => false
-gem "rails",                            "~>7.1.5", ">=7.1.5.1"
+gem "rails",                            "~>7.2.0"
```
```suggestion
gem "rails",                            "~>7.2.2", ">= 7.2.2.1"
```

Fryguy commented:
@jrafanie When you get a chance, can you rebase? I want to see if I fixed the bot and this PR happens to trigger the bot problem.

jrafanie commented:
> @jrafanie When you get a chance, can you rebase? I want to see if I fixed the bot and this PR happens to trigger the bot problem.

rebased @Fryguy

jrafanie commented on config/application.rb:
```diff
@@ -110,6 +110,8 @@ class Application < Rails::Application
 
     config.autoload_paths += config.eager_load_paths
 
+    config.paths["config/secrets"] = []
```
Done

Fryguy commented on lib/vmdb/initializer.rb:
```diff
@@ -15,11 +15,24 @@ def self.init
+      return if base == db_token
 
-      Rails.application.config.secret_key_base = token
+      new_token = if db_token.present?
+                    # use db_token above all
+                    db_token
+                  else
+                    # generate a new token if base isn't already configured
+                    return if /^\h{128}$/.match?(base)
+
+                    SecureRandom.hex(64)
+                  end
```
I don't understand this code relative to what was there before. It's like basically the same but not?  Can we discuss?

agrare commented on lib/vmdb/initializer.rb:
```diff
@@ -15,11 +15,24 @@ def self.init
+                    db_token
+                  else
+                    # generate a new token if base isn't already configured
+                    return if /^\h{128}$/.match?(base)
```
Would it make sense to move all of the returns together?  Maybe `return if base == db_token || /^\h{128}$/.match?(base)` ?

Fryguy commented on bin/test_database_connectivity:
```diff
@@ -0,0 +1,8 @@
+
+bundle exec rake test:verify_no_db_access_loading_rails_environment
+bundle exec rake test:verify_connectable_with_valid_database
+bundle exec rake test:verify_not_connectable_with_invalid_database
```
The `ActiveRecord::DatabaseConnectionError` was needed because we were connecting to what is not an invalid hostname in the dummy configuation. Previously /var/lib/postgresql was the socket on Travis, but since we're not on Travis we don't use the socket anymore.  It's still worth keeping the constant in the rescue, but not in the dummy configuration.

jrafanie replied:
I think you might have an older provider checked out...

```
commit 9af18f5203cd6e1553c0af2a775395d05a094de2
Author: Jason Frey <fryguy9@gmail.com>
Date:   Tue Jul 1 10:54:55 2025 -0400

    Update provider generator to use VcrSecrets

diff --git a/lib/generators/manageiq/plugin/templates/lib/%plugin_path%/engine.rb b/lib/generators/manageiq/plugin/templates/lib/%plugin_path%/engine.rb
index 2915c75355..0e4d523b2d 100644
--- a/lib/generators/manageiq/plugin/templates/lib/%plugin_path%/engine.rb
+++ b/lib/generators/manageiq/plugin/templates/lib/%plugin_path%/engine.rb
@@ -4,11 +4,6 @@

     config.autoload_paths << root.join('lib').to_s

-    initializer :append_secrets do |app|
-      app.config.paths["config/secrets"] << root.join("config", "secrets.defaults.yml").to_s
-      app.config.paths["config/secrets"] << root.join("config", "secrets.yml").to_s
-    end
-
     def self.vmdb_plugin?
       true
     end
```

Fryguy commented:
@jrafanie Please run cross-repo and cypress one more time before merge.

jrafanie commented:
Cross repo is still 💚, cypress is passing locally

Fryguy commented:
Great work @jrafanie and @kbrock !! 🎉
